Update systemd-cryptenroll to enroll TPM using only public key

ddstreet opened the pull request:
This adds support for calculating a sealed object using only a parent public key (and its location). By default, the parent public key is assumed to be a SRK, located in the standard SRK address of 0x81000001.

This change allows enrolling a TPM that the current system doesn't have access to.

This is on top of #28398 and should wait for 255 to open.

<!-- devel-freezer = {"comment-id":"1650518185","freezing-tag":"v255-rc1"} -->

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -387,6 +387,28 @@
+        <term><option>--tpm2-external-key=</option><replaceable>PATH</replaceable></term>
+
+        <listitem><para>Enroll an external TPM2 security chip. Expects a path referring to the TPM2 public
+        key in TPM2B_PUBLIC format. This cannot be used with <option>--tpm2-device</option>, and this does
```
why in TPM2B_PUBLIC format? we so far are always using PEM format when interfacing with the user. Why use TPM2B_PUBLIC here? why not convert on the fly? all these tpm concepts are hard to grok anyway, i'd rather not confuse users further by pushing two different key marshallings on them.

please suffix `--tpm2-device=` with `=`

ddstreet replied:
> why in TPM2B_PUBLIC format? we so far are always using PEM format when interfacing with the user. Why use TPM2B_PUBLIC here? why not convert on the fly? all these tpm concepts are hard to grok anyway, i'd rather not confuse users further by pushing two different key marshallings on them.

we do need more than just the RSA/ECC key data for the calculations, we also need the name hash alg (e.g. SHA256) as well as the symmetric alg (e.g. AES-128-CFB). And finally, since we now store the serialized SRK, we must have the entire thing, since that (plus its "name" and handle index) is what goes into the serialized object.

> 
> please suffix `--tpm2-device=` with `=`

ack, updated

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -387,6 +387,28 @@
+        configuration in <filename>/etc/crypttab</filename> as would be used for direct TPM
+        enrollment.</para>
+
+        <para>Also see <option>--tpm2-external-location</option>, which is the TPM handle location where the
```
similar, please suffix the option with `=`

ddstreet replied:
ack

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -387,6 +387,28 @@
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>--tpm2-external-location=</option><replaceable>LOCATION</replaceable></term>
```
"location", is that the right terminology? isn't he right name for this thing "handle"?

ddstreet replied:
yeah it's not actually a 'location' but 'handle' seems confusing. i updated it to `--tpm2-handle-index` which i hope is a better way to refer to it? The TPM spec refers to the NV range as 'NV indexes' so maybe it makes sense to refer to all handles as 'handle indexes'. Just referring to it as 'handle' seems confusing since we already have the Tpm2Handle object, and the EAPI's ESYS_TR handle, and then this is the low-level TPM_HANDLE. Hard for me to get a handle on all the handles. :)

I also removed the 'external' part from the index, as specifying the handle index could apply to a local tpm as well.

Technically, we could leave this parameter out compeltely, and just always use the SRK and not give end users the option to use any other TPM handle (index). However it doesn't add any complexity (or at least, not any significant complexity), and I think it could be useful to some people.

ddstreet replied:
on thinking about this a bit, i changed it as you suggested to `--tpm2-handle=`. I think at least from the user's perspective, there shouldn't be any confusion.

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -387,6 +387,28 @@
+      <varlistentry>
+        <term><option>--tpm2-external-location=</option><replaceable>LOCATION</replaceable></term>
+
+        <listitem><para>When used with <option>--tpm2-external-key</option>, specifies the external TPM handle location of the public key. This default to the Storage Root Key (SRK) defined location <literal>0x81000001</literal>.</para></listitem>
```
suffix with `=`

ddstreet replied:
ack

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -387,6 +387,28 @@
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--tpm2-external-key=</option><replaceable>PATH</replaceable></term>
```
so this explanation here does not mention the word SRK at all, but this is almost certainly going to be the SRK of the target machine, no? Maybe say it like that.

ddstreet replied:
yep, i reworded this to clarify it's the srk.

poettering commented on src/cryptenroll/cryptenroll.c:
```diff
@@ -358,6 +369,26 @@ static int parse_argv(int argc, char *argv[]) {
                 }
 
+                case ARG_TPM2_EXTERNAL_LOCATION:
+                        r = safe_atou32(optarg, &arg_tpm2_external_location);
```
are all handles valid? i.e. handle 0 too? should we refuse some invalid handles here?

poettering replied:
hmm shouldn't this always be parsed in hex? even if unprefixed with `0x`?

ddstreet replied:
yes! updated, thanks

ddstreet commented:
last push was just a rebase on main, no changes for the review yet

ddstreet commented:
rebased on #28916, now starting to go over review comments

ddstreet commented:
ok i think this is ready for re-review, PTAL; I'm also going to go over each commit once more on Monday.

bluca commented:
please rebase and add `<xi:include href="version-info.xml" xpointer="v255"/>` in the manpage paragraphs of the new options (it's a new thing)

ddstreet commented:
rebased and added new `<xi:...>` tags to man pages additions.

ddstreet commented:
rebased on main

bluca commented on src/shared/openssl-util.c:
```diff
@@ -858,6 +858,56 @@ int ecc_pkey_new(int curve_id, EVP_PKEY **ret) {
+        assert(ret_shared_secret);
+        assert(ret_shared_secret_size);
+
+        r = ecc_pkey_to_curve_x_y(peerkey, &curve_id, NULL, NULL, NULL, NULL);
```
name the null parameters with a comment `/* foobar= */ NULL`, etc

bluca commented:
```
19:55:25 #3  0x00007ff3f09dc007 in log_assert_failed (text=text@entry=0x55b0c64f9860 "tpm2_unseal( NULL, 0, 0, NULL, 0, 0, NULL, NULL, 0, blob, blob_size, NULL, 0, serialized_parent, serialized_parent_size, &unsealed_secret, &unsealed_secret_size) >= 0", file=file@entry=0x55b0c64f301b "src/test/test-tpm2.c", line=line@entry=1031, func=func@entry=0x55b0c64fe900 <__func__.12> "calculate_seal_and_unseal") at ../build/src/basic/log.c:947
19:55:25 #4  0x000055b0c64e10e0 in calculate_seal_and_unseal (parent_index=2164260863, parent_public=0x55b0c7f97ce0) at ../build/src/test/test-tpm2.c:1031
19:55:25 #5  0x000055b0c64e12e0 in check_calculate_seal (c=0x55b0c7f29d60) at ../build/src/test/test-tpm2.c:1086
19:55:25 #6  0x000055b0c64e149a in test_tpm_required_tests () at ../build/src/test/test-tpm2.c:1104
19:55:25 #7  0x000055b0c64f2274 in run_test_table () at ../build/src/shared/tests.h:123
19:55:25 #8  0x000055b0c64f23d3 in main (argc=1, argv=0x7ffcab0cbc48) at ../build/src/test/test-tpm2.c:1372
19:55:25 (gdb) #0  0x00007ff3efe8e83c in ?? () from /usr/lib/libc.so.6
19:55:25 No symbol table info available.
19:55:25 #1  0x00007ff3efe3e668 in raise () from /usr/lib/libc.so.6
19:55:25 No symbol table info available.
19:55:25 #2  0x00007ff3efe264b8 in abort () from /usr/lib/libc.so.6
19:55:25 No symbol table info available.
19:55:25 #3  0x00007ff3f09dc007 in log_assert_failed (
19:55:25     text=text@entry=0x55b0c64f9860 "tpm2_unseal( NULL, 0, 0, NULL, 0, 0, NULL, NULL, 0, blob, blob_size, NULL, 0, serialized_parent, serialized_parent_size, &unsealed_secret, &unsealed_secret_size) >= 0", 
19:55:25     file=file@entry=0x55b0c64f301b "src/test/test-tpm2.c", 
19:55:25     line=line@entry=1031, 
19:55:25     func=func@entry=0x55b0c64fe900 <__func__.12> "calculate_seal_and_unseal")
19:55:25     at ../build/src/basic/log.c:947
19:55:25 No locals.
19:55:25 #4  0x000055b0c64e10e0 in calculate_seal_and_unseal (parent_index=2164260863, 
19:55:25     parent_public=0x55b0c7f97ce0) at ../build/src/test/test-tpm2.c:1031
19:55:25         secret_string = 0x55b0c7f89ea0 "The classified documents are in room 80ffffff"
19:55:25         __func__ = "calculate_seal_and_unseal"
19:55:25         secret_size = 46
19:55:25         blob = 0x55b0c7fa28e0
19:55:25         blob_size = 462
19:55:25         serialized_parent = 0x55b0c7f98200
19:55:25         serialized_parent_size = 328
19:55:25         unsealed_secret = 0x0
19:55:25         unsealed_secret_size = 0
19:55:25         unsealed_string = <optimized out>
19:55:25 #5  0x000055b0c64e12e0 in check_calculate_seal (c=0x55b0c7f29d60)
19:55:25     at ../build/src/test/test-tpm2.c:1086
19:55:25         template = {size = 612, publicArea = {type = 1, nameAlg = 11, 
19:55:25             objectAttributes = 197746, authPolicy = {size = 0, 
19:55:25               buffer = '\000' <repeats 63 times>}, parameters = {
19:55:25               keyedHashDetail = {scheme = {scheme = 6, details = {hmac = {
19:55:25                       hashAlg = 128}, exclusiveOr = {hashAlg = 128, 
19:55:25                       kdf = 67}, null = {empty = "\200"}}}}, symDetail = {
19:55:25                 sym = {algorithm = 6, keyBits = {aes = 128, sm4 = 128, 
19:55:25                     camellia = 128, sym = 128, exclusiveOr = 128, null = {
19:55:25                       empty = "\200"}}, mode = {aes = 67, sm4 = 67, 
19:55:25                     camellia = 67, sym = 67, exclusiveOr = {empty = "C"}, 
19:55:25                     null = {empty = "C"}}}}, rsaDetail = {symmetric = {
19:55:25                   algorithm = 6, keyBits = {aes = 128, sm4 = 128, 
19:55:25                     camellia = 128, sym = 128, exclusiveOr = 128, null = {
19:55:25                       empty = "\200"}}, mode = {aes = 67, sm4 = 67, 
19:55:25                     camellia = 67, sym = 67, exclusiveOr = {empty = "C"}, 
19:55:25                     null = {empty = "C"}}}, scheme = {scheme = 16, details = {
19:55:25                     ecdh = {hashAlg = 0}, ecmqv = {hashAlg = 0}, rsassa = {
19:55:25                       hashAlg = 0}, rsapss = {hashAlg = 0}, ecdsa = {
19:55:25                       hashAlg = 0}, ecdaa = {hashAlg = 0, count = 0}, sm2 = {
19:55:25                       hashAlg = 0}, ecschnorr = {hashAlg = 0}, rsaes = {
19:55:25                       empty = ""}, oaep = {hashAlg = 0}, anySig = {
19:55:25                       hashAlg = 0}, null = {empty = ""}}}, keyBits = 2048, 
19:55:25                 exponent = 0}, eccDetail = {symmetric = {algorithm = 6, 
19:55:25                   keyBits = {aes = 128, sm4 = 128, camellia = 128, sym = 128, 
19:55:25                     exclusiveOr = 128, null = {empty = "\200"}}, mode = {
19:55:25                     aes = 67, sm4 = 67, camellia = 67, sym = 67, 
19:55:25                     exclusiveOr = {empty = "C"}, null = {empty = "C"}}}, 
19:55:25                 scheme = {scheme = 16, details = {ecdh = {hashAlg = 0}, 
19:55:25                     ecmqv = {hashAlg = 0}, rsassa = {hashAlg = 0}, rsapss = {
19:55:25                       hashAlg = 0}, ecdsa = {hashAlg = 0}, ecdaa = {
19:55:25                       hashAlg = 0, count = 0}, sm2 = {hashAlg = 0}, 
19:55:25                     ecschnorr = {hashAlg = 0}, rsaes = {empty = ""}, oaep = {
19:55:25                       hashAlg = 0}, anySig = {hashAlg = 0}, null = {
19:55:25                       empty = ""}}}, curveID = 2048, kdf = {scheme = 0, 
19:55:25                   details = {mgf1 = {hashAlg = 0}, kdf1_sp800_56a = {
19:55:25                       hashAlg = 0}, kdf2 = {hashAlg = 0}, kdf1_sp800_108 = {
19:55:25                       hashAlg = 0}, null = {empty = ""}}}}, asymDetail = {
19:55:25                 symmetric = {algorithm = 6, keyBits = {aes = 128, sm4 = 128, 
19:55:25                     camellia = 128, sym = 128, exclusiveOr = 128, null = {
19:55:25                       empty = "\200"}}, mode = {aes = 67, sm4 = 67, 
19:55:25                     camellia = 67, sym = 67, exclusiveOr = {empty = "C"}, 
19:55:25                     null = {empty = "C"}}}, scheme = {scheme = 16, details = {
19:55:25                     ecdh = {hashAlg = 0}, ecmqv = {hashAlg = 0}, rsassa = {
19:55:25                       hashAlg = 0}, rsapss = {hashAlg = 0}, ecdsa = {
19:55:25                       hashAlg = 0}, ecdaa = {hashAlg = 0, count = 0}, sm2 = {
19:55:25                       hashAlg = 0}, ecschnorr = {hashAlg = 0}, rsaes = {
19:55:25                       empty = ""}, oaep = {hashAlg = 0}, anySig = {
19:55:25                       hashAlg = 0}, null = {empty = ""}}}}}, unique = {
19:55:25               keyedHash = {size = 0, buffer = '\000' <repeats 63 times>}, 
19:55:25               sym = {size = 0, buffer = '\000' <repeats 63 times>}, rsa = {
19:55:25                 size = 0, buffer = '\000' <repeats 511 times>}, ecc = {x = {
19:55:25                   size = 0, buffer = '\000' <repeats 127 times>}, y = {
19:55:25                   size = 0, buffer = '\000' <repeats 127 times>}}, derive = {
19:55:25                 label = {size = 0, buffer = '\000' <repeats 31 times>}, 
19:55:25                 context = {size = 0, buffer = '\000' <repeats 31 times>}}}}}
19:55:25         handle = 0x55b0c7f92ee0
19:55:25         alg = <optimized out>
19:55:25         public = 0x55b0c7f97ce0
19:55:25         index = 2164260863
19:55:25         i = 0
19:55:25         __func__ = "check_calculate_seal"
19:55:25         r = <optimized out>
19:55:25         srk_public = 0x55b0c7f304a0
19:55:25         test_algs = {1, 35}
19:55:25 #6  0x000055b0c64e149a in test_tpm_required_tests ()
19:55:25     at ../build/src/test/test-tpm2.c:1104
19:55:25         c = 0x55b0c7f29d60
19:55:25 #7  0x000055b0c64f2274 in run_test_table () at ../build/src/shared/tests.h:123
19:55:25         r2 = <optimized out>
19:55:25         t = 0x55b0c65020a0 <__unique_prefix_static_test_table_entry30>
19:55:25         tests = 0x0
19:55:25         r = 0
19:55:25         ran = true
19:55:25         e = <optimized out>
19:55:25         __func__ = "run_test_table"
19:55:25 #8  0x000055b0c64f23d3 in main (argc=1, argv=0x7ffcab0cbc48)
19:55:25     at ../build/src/test/test-tpm2.c:1372
19:55:25         _intro = 0x0
19:55:25         _outro = 0x0
19:55:25         _r = 0
19:55:25         _q = 0
```

ddstreet commented on src/shared/openssl-util.c:
```diff
@@ -134,6 +134,88 @@ int openssl_digest_many(
+                size_t n_data,
+                void **ret_digest,
+                size_t *ret_digest_size) {
+
```
ack, updated

ddstreet commented on src/shared/openssl-util.c:
```diff
@@ -216,6 +216,85 @@ int openssl_hmac_many(
+#if OPENSSL_VERSION_MAJOR >= 3
+        assert(mode);
+        assert(digest);
+
```
ack, updated

bluca commented:
```
23:48:35 #3  0x00007ffb173dc202 in log_assert_failed (
23:48:35     text=text@entry=0x562215c08860 "tpm2_unseal( NULL, 0, 0, NULL, 0, 0, NULL, NULL, 0, blob, blob_size, NULL, 0, serialized_parent, serialized_parent_size, &unsealed_secret, &unsealed_secret_size) >= 0", 
23:48:35     file=file@entry=0x562215c0201b "src/test/test-tpm2.c", 
23:48:35     line=line@entry=1031, 
23:48:35     func=func@entry=0x562215c0d900 <__func__.12> "calculate_seal_and_unseal")
23:48:35     at ../build/src/basic/log.c:947
23:48:35 No locals.
23:48:35 #4  0x0000562215bf00e0 in calculate_seal_and_unseal (parent_index=2164260863, 
23:48:35     parent_public=0x56221735ece0) at ../build/src/test/test-tpm2.c:1031
23:48:35         secret_string = 0x562217350ea0 "The classified documents are in room 80ffffff"
23:48:35         __func__ = "calculate_seal_and_unseal"
23:48:35         secret_size = 46
23:48:35         blob = 0x5622173698e0
23:48:35         blob_size = 462
23:48:35         serialized_parent = 0x56221735f200
23:48:35         serialized_parent_size = 328
23:48:35         unsealed_secret = 0x0
23:48:35         unsealed_secret_size = 0
23:48:35         unsealed_string = <optimized out>
23:48:35 #5  0x0000562215bf02e0 in check_calculate_seal (c=0x5622172f0d60)
23:48:35     at ../build/src/test/test-tpm2.c:1086
23:48:35         template = {size = 612, publicArea = {type = 1, nameAlg = 11, 
23:48:35             objectAttributes = 197746, authPolicy = {size = 0, 
23:48:35               buffer = '\000' <repeats 63 times>}, parameters = {
23:48:35               keyedHashDetail = {scheme = {scheme = 6, details = {hmac = {
23:48:35                       hashAlg = 128}, exclusiveOr = {hashAlg = 128, 
23:48:35                       kdf = 67}, null = {empty = "\200"}}}}, symDetail = {
23:48:35                 sym = {algorithm = 6, keyBits = {aes = 128, sm4 = 128, 
23:48:35                     camellia = 128, sym = 128, exclusiveOr = 128, null = {
23:48:35                       empty = "\200"}}, mode = {aes = 67, sm4 = 67, 
23:48:35                     camellia = 67, sym = 67, exclusiveOr = {empty = "C"}, 
23:48:35                     null = {empty = "C"}}}}, rsaDetail = {symmetric = {
23:48:35                   algorithm = 6, keyBits = {aes = 128, sm4 = 128, 
23:48:35                     camellia = 128, sym = 128, exclusiveOr = 128, null = {
23:48:35                       empty = "\200"}}, mode = {aes = 67, sm4 = 67, 
23:48:35                     camellia = 67, sym = 67, exclusiveOr = {empty = "C"}, 
23:48:35                     null = {empty = "C"}}}, scheme = {scheme = 16, details = {
23:48:35                     ecdh = {hashAlg = 0}, ecmqv = {hashAlg = 0}, rsassa = {
23:48:35                       hashAlg = 0}, rsapss = {hashAlg = 0}, ecdsa = {
23:48:35                       hashAlg = 0}, ecdaa = {hashAlg = 0, count = 0}, sm2 = {
23:48:35                       hashAlg = 0}, ecschnorr = {hashAlg = 0}, rsaes = {
23:48:35                       empty = ""}, oaep = {hashAlg = 0}, anySig = {
23:48:35                       hashAlg = 0}, null = {empty = ""}}}, keyBits = 2048, 
23:48:35                 exponent = 0}, eccDetail = {symmetric = {algorithm = 6, 
23:48:35                   keyBits = {aes = 128, sm4 = 128, camellia = 128, sym = 128, 
23:48:35                     exclusiveOr = 128, null = {empty = "\200"}}, mode = {
23:48:35                     aes = 67, sm4 = 67, camellia = 67, sym = 67, 
23:48:35                     exclusiveOr = {empty = "C"}, null = {empty = "C"}}}, 
23:48:35                 scheme = {scheme = 16, details = {ecdh = {hashAlg = 0}, 
23:48:35                     ecmqv = {hashAlg = 0}, rsassa = {hashAlg = 0}, rsapss = {
23:48:35                       hashAlg = 0}, ecdsa = {hashAlg = 0}, ecdaa = {
23:48:35                       hashAlg = 0, count = 0}, sm2 = {hashAlg = 0}, 
23:48:35                     ecschnorr = {hashAlg = 0}, rsaes = {empty = ""}, oaep = {
23:48:35                       hashAlg = 0}, anySig = {hashAlg = 0}, null = {
23:48:35                       empty = ""}}}, curveID = 2048, kdf = {scheme = 0, 
23:48:35                   details = {mgf1 = {hashAlg = 0}, kdf1_sp800_56a = {
23:48:35                       hashAlg = 0}, kdf2 = {hashAlg = 0}, kdf1_sp800_108 = {
23:48:35                       hashAlg = 0}, null = {empty = ""}}}}, asymDetail = {
23:48:35                 symmetric = {algorithm = 6, keyBits = {aes = 128, sm4 = 128, 
23:48:35                     camellia = 128, sym = 128, exclusiveOr = 128, null = {
23:48:35                       empty = "\200"}}, mode = {aes = 67, sm4 = 67, 
23:48:35                     camellia = 67, sym = 67, exclusiveOr = {empty = "C"}, 
23:48:35                     null = {empty = "C"}}}, scheme = {scheme = 16, details = {
23:48:35                     ecdh = {hashAlg = 0}, ecmqv = {hashAlg = 0}, rsassa = {
23:48:35                       hashAlg = 0}, rsapss = {hashAlg = 0}, ecdsa = {
23:48:35                       hashAlg = 0}, ecdaa = {hashAlg = 0, count = 0}, sm2 = {
23:48:35                       hashAlg = 0}, ecschnorr = {hashAlg = 0}, rsaes = {
23:48:35                       empty = ""}, oaep = {hashAlg = 0}, anySig = {
23:48:35                       hashAlg = 0}, null = {empty = ""}}}}}, unique = {
23:48:35               keyedHash = {size = 0, buffer = '\000' <repeats 63 times>}, 
23:48:35               sym = {size = 0, buffer = '\000' <repeats 63 times>}, rsa = {
23:48:35                 size = 0, buffer = '\000' <repeats 511 times>}, ecc = {x = {
23:48:35                   size = 0, buffer = '\000' <repeats 127 times>}, y = {
23:48:35                   size = 0, buffer = '\000' <repeats 127 times>}}, derive = {
23:48:35                 label = {size = 0, buffer = '\000' <repeats 31 times>}, 
23:48:35                 context = {size = 0, buffer = '\000' <repeats 31 times>}}}}}
23:48:35         handle = 0x562217359ee0
23:48:35         alg = <optimized out>
23:48:35         public = 0x56221735ece0
23:48:35         index = 2164260863
23:48:35         i = 0
23:48:35         __func__ = "check_calculate_seal"
23:48:35         r = <optimized out>
23:48:35         srk_public = 0x5622172f74a0
23:48:35         test_algs = {1, 35}
23:48:35 #6  0x0000562215bf049a in test_tpm_required_tests ()
23:48:35     at ../build/src/test/test-tpm2.c:1104
23:48:35         c = 0x5622172f0d60
23:48:35 #7  0x0000562215c01274 in run_test_table () at ../build/src/shared/tests.h:123
23:48:35         r2 = <optimized out>
23:48:35         t = 0x562215c110a0 <__unique_prefix_static_test_table_entry30>
23:48:35         tests = 0x0
23:48:35         r = 0
23:48:35         ran = true
23:48:35         e = <optimized out>
23:48:35         __func__ = "run_test_table"
23:48:35 #8  0x0000562215c013d3 in main (argc=1, argv=0x7ffd7fc1cd28)
23:48:35     at ../build/src/test/test-tpm2.c:1372
23:48:35         _intro = 0x0
23:48:35         _outro = 0x0
23:48:35         _r = 0
23:48:35         _q = 0
```

poettering commented on src/shared/openssl-util.c:
```diff
@@ -223,6 +223,105 @@ int openssl_hmac_many(
+
+        _cleanup_free_ char *cipher_alg = NULL;
+        r = asprintf(&cipher_alg, "%s-%zu-%s", alg, bits, mode);
+        if (r < 0)
```
I'd shorten this to:

```c
if (asprintf(…) < 0)
        …
```

ddstreet replied:
ack

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3749,6 +3749,73 @@ int tpm2_tpm2b_public_from_pem(const void *pem, size_t pem_size, TPM2B_PUBLIC *r
 #endif
 }
 
+int tpm2_create_blob(
```
uh, this could use a better name: tpm2_marshal_keypair() or something?

poettering replied:
ah, you are extending this later.

maybe tpm2_marshal_structs() then?

poettering replied:
sure.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3749,6 +3749,73 @@ int tpm2_tpm2b_public_from_pem(const void *pem, size_t pem_size, TPM2B_PUBLIC *r
+        return 0;
+}
+
+int tpm2_extract_blob(
```
tpm2_unmarshal_keypair()?

ddstreet replied:
as above; i changed this to tpm2_unmarshal_blob()

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3763,6 +3812,8 @@ int tpm2_create_blob(
 
         size_t max_size = sizeof(*private) + sizeof(*public);
+        if (seed)
+                max_size += sizeof(*seed);
```
maybe make all three parts optional then? seems weird only allowing one of the three to be optional

ddstreet replied:
that won't work. this 'blob' is not a standard format, so if all 3 parts were optional there would be no sane way to tell what each part was. Additionally, since this is supposed to store the data for a sealed object, the public and private parts can't be optional - those are always required for a sealed object.

poettering replied:
indeed

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1105,18 +1105,11 @@ static int tpm2_get_best_srk_template(Tpm2Context *c, TPMT_PUBLIC *ret_template)
-static int tpm2_get_srk(
+/* Get the Tpm2Handle at the requested index. Returns 1 if found, 0 if the index is empty, or < 0 on
+ * error. Also see tpm2_get_srk() below; the SRK is a commonly used persistent Tpm2Handle. */
+int tpm2_get_handle(
```
maybe be explicit? tpm2_get_handle_from_index()?

ddstreet replied:
Since the `tpm2_get_handle()` and `tpm2_get_handle_index()` functions are reciprocal functions, I changed the names to `tpm2_index_from_handle()` and `tpm2_index_to_handle()`, how does that sound?

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -4003,6 +4065,644 @@ static int tpm2_deserialize(
+        assert(ret_key);
+        assert(ret_key_len);
+
+        size_t len = (bits + 7) / 8;
```
DIV_ROUND_UP(bits, 8);

ddstreet replied:
ack, thanks

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -4003,6 +4065,644 @@ static int tpm2_deserialize(
+         * CLEAR. The unused bits of the most significant octet (MSO) are masked off and not shifted." */
+        size_t partial = bits % 8;
+        if (partial > 0)
+                ((uint8_t*) buf)[0] &= 0xff >> (8 - partial);
```
i find shifting signed values confusing. maybe make this 0xffu to be explicit?

ddstreet replied:
good point, updated, thanks

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -4003,6 +4065,644 @@ static int tpm2_deserialize(
+        memcpy(&info[info_len], context_u, context_u_size);
+        info_len += context_u_size;
+        memcpy(&info[info_len], context_v, context_v_size);
+        info_len += context_v_size;
```
maybe instead:

```c
info_len = mempcpy(mempcpy(stpcpy(info, label) + 1, 
                                  context_u, context_u_size),
                                  context_v, context_vsize))) - info;
```

you get the idea

ddstreet replied:
yep thanks, and i added an assert to verify the amount copied was correct

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1124,6 +1131,28 @@ int tpm2_get_best_srk_template(Tpm2Context *c, TPMT_PUBLIC *ret_template) {
 }
 
+/* Get the TPM2_HANDLE index for the provided Tpm2Handle. */
+int tpm2_get_handle_index(Tpm2Context *c, const Tpm2Handle *handle, TPM2_HANDLE *ret_index) {
```
tpm2_get_index_from_handle()?

ddstreet replied:
updated, leaving the `_get` part out

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -69,28 +73,40 @@ static TSS2_RC (*sym_Esys_TRSess_SetAttributes)(ESYS_CONTEXT *esysContext, ESYS_
-static TSS2_RC (*sym_Tss2_MU_UINT32_Marshal)(UINT32 src, uint8_t buffer[], size_t buffer_size, size_t *offset) = NULL;
-
-static const char* (*sym_Tss2_RC_Decode)(TSS2_RC rc) = NULL;
+#define DEFINE_MU(TYPE)                                                 \
```
maybe DEFINE_MU_SYM()?

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -69,28 +73,40 @@ static TSS2_RC (*sym_Esys_TRSess_SetAttributes)(ESYS_CONTEXT *esysContext, ESYS_
-static const char* (*sym_Tss2_RC_Decode)(TSS2_RC rc) = NULL;
+#define DEFINE_MU(TYPE)                                                 \
+        TSS2_RC (*sym_Tss2_MU_##TYPE##_Marshal)(TYPE const *src, uint8_t buffer[], size_t buffer_size, size_t *offset) = NULL; \
+        TSS2_RC (*sym_Tss2_MU_##TYPE##_Unmarshal)(uint8_t const buffer[], size_t buffer_size, size_t *offset, TYPE *dest) = NULL
```
hmm, another approach might be to add a macro do dlfcn-util.h that goes something like this:

```c
DSLYM_DEFINE(func) \
        typeof(func) sym_##func = NULL
```

And then we stop specifying the full type each time, and instead can just do:

```c
DLSYM_DEFINE(Tss2_MU_UIN32_Marshal);
…
```

and so on. Any maybe prefix a "static"....

poettering commented on src/shared/tpm2-util.h:
```diff
@@ -50,6 +50,16 @@ static inline bool TPM2_PCR_MASK_VALID(uint32_t pcr_mask) {
+/* Calls dlopen_tpm2() and returns 0 on success, < 0 on error. Also logs on error. Useful in "elvis operator"
+ * ternary checks. */
+static inline int tpm2_dlopen(void) {
+        int r = dlopen_tpm2();
```
jmmpf. this is just very very confusing. the difference is so not obvious...

maybe instead just make the log level configurable, dlopen_many_sym_or_warn() accepts that too anyway

poettering commented on src/shared/tpm2-util.h:
```diff
@@ -259,6 +269,173 @@ int tpm2_tpm2b_public_to_fingerprint(const TPM2B_PUBLIC *public, void **ret_fing
+                        : log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE), \
+                                          "Failed to unmarshal %s: %s", \
+                                          desc, sym_Tss2_RC_Decode(UNIQ_T(RC, u))); \
+        })
```
hmpff, these "functions" are conceptually low-level "library-like" basics and we said that "library-like" functions should not log on their own. but here you do...

mrc0mmand replied:
I wonder if it wouldn't be better to use functions instead of macros here, since the macros are very "function-like" and unnecessarily cryptic.

poettering commented on src/shared/tpm2-util.h:
```diff
@@ -259,6 +269,173 @@ int tpm2_tpm2b_public_to_fingerprint(const TPM2B_PUBLIC *public, void **ret_fing
+                int UNIQ_T(R, u) = read_full_file(f, &UNIQ_T(B, u), &UNIQ_T(S, u)); \
+                if (UNIQ_T(R, u) >= 0)                                  \
+                        UNIQ_T(R, u) = _tpm2_unmarshal(desc, dst, (uint8_t*) UNIQ_T(B, u), UNIQ_T(S, u), 0, ret_size, CONCATENATE(u, _tpm2_unmarshal_from_file)); \
+                UNIQ_T(R, u);                                           \
```
io errors will not be logged here, but marshalling errors will be.... we usually want that either all errors are logged or none

ddstreet commented on src/shared/openssl-util.c:
```diff
@@ -223,6 +223,105 @@ int openssl_hmac_many(
+
+        /* Verify enough key data was provided. */
+        int cipher_key_length = EVP_CIPHER_key_length(cipher);
+        if ((size_t) cipher_key_length > key_size)
```
yep, added

ddstreet commented on src/shared/openssl-util.c:
```diff
@@ -223,6 +223,105 @@ int openssl_hmac_many(
+        int cipher_iv_length = EVP_CIPHER_iv_length(cipher);
+        _cleanup_free_ void *zero_iv = NULL;
+        if (!iv) {
+                zero_iv = malloc0(cipher_iv_length);
```
ack

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -409,6 +409,16 @@
+
+        <listitem><para>Configures which parent key to use for sealing, using the TPM handle of the key. The
+        default is the Storage Root Key (SRK) handle index <literal>0x81000001</literal>. This should not be
+        changed unless you know what you are doing.</para>
```
should document expected range and syntax

ddstreet replied:
yep, updated the wording, thanks

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -409,11 +409,34 @@
+      <varlistentry>
+        <term><option>--tpm2-external-key=</option><replaceable>PATH</replaceable></term>
+
+        <listitem><para>Enroll an external TPM2 security chip. Expects a path referring to the TPM2 public
```
maybe say what "external" is supposed to mean. Maybe should use words like "offline"?

poettering replied:
is the word "external" the terminology used elsewhere for this concept?

otherwise maybe call this an "offline key" or "disconnected key" or so?

ddstreet replied:
Good point, how about calling the parameter `--tpm2-device-key=` so it matches up with the `--tpm2-device=` parameter? I reworded the man page also to indicate it performs the same operation as `--tpm2-device=` but using calculation, instead of direct access to the tpm.

poettering commented:
looks good. (the marshalling macros are a bit too magic for me I must say though)

ddstreet commented:
ok pushed to address the CI failure (added https://github.com/systemd/systemd/pull/28519/commits/136dc1cfdc070610be98d09a03337190fa0206dc for that) and addressed some review comments but not all; will continue tomorrow

ddstreet commented:
I would change the 'please-review' tag back to 'needs-rework' but I don't think I'm able to adjust the tags

ddstreet commented on src/cryptenroll/cryptenroll.c:
```diff
@@ -358,6 +363,13 @@ static int parse_argv(int argc, char *argv[]) {
+                case ARG_TPM2_HANDLE:
+                        r = safe_atou32_full(optarg, 16, &arg_tpm2_handle);
+                        if (r < 0)
+                                return r;
```
ack

ddstreet commented:
ok i think i addressed all the review comments *except* those for the marshalling commit; I'll work on that now

ddstreet commented:
I split out the openssl functions into #29183 since this PR has got rather large.

ddstreet commented:
moving back to draft for now

bluca commented:
@ddstreet is this ready for un-draft, after a rebase?

poettering commented:
ping?

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -4338,8 +4412,67 @@ static int tpm2_unmarshal_blob(
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                               "Failed to unmarshal encrypted seed: %s", sym_Tss2_RC_Decode(rc));
+        }
+
```
hmm, actually, we should probably add a check after this that verifies that offset == blob_size in the end, as there should not be trailing garbage. but maybe that's for another time

ddstreet commented:
> ping?

really sorry for all the delays.

I think this is ready for final review; a lot of the commits were split off into smaller PRs and already merged, so this is now mainly just the commit that adds the sealing calculation functions.

I have *not* tested this with the new pcrlock functionality, but it should work as the pcrlock logic adjusts the calculated authorization policy which this doesn't touch.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -4403,6 +4536,632 @@ int tpm2_deserialize(
+
+        size_t hash_size = tpm2_hash_alg_to_size(parent->publicArea.nameAlg);
+        if (hash_size == 0)
+                return -EOPNOTSUPP;
```
maybe `log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "…` here?

poettering replied:
similar for similar cases, but also doesn't really matter

ddstreet replied:
my thinking here was that `tpm2_hash_alg_to_size()` would have just logged to debug in the case of an unknown alg, so it probably wasn't needed to log to debug again?

However I did mess up the handling of the return value, as `tpm2_hash_alg_to_size()` returns < 0 on error (not 0), oops...i will fix that

poettering commented:
There's some memory issue:

```
23:11:28 [   99.210399] testsuite-70.sh[4481]: ==4481==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x000000000000 bp 0x7ffe228cc1f0 sp 0x7ffe228cbfb8 T0)
23:11:28 [   99.211414] testsuite-70.sh[4481]: ==4481==Hint: pc points to the zero page.
23:11:28 [   99.212107] testsuite-70.sh[4481]: ==4481==The signal is caused by a READ memory access.
23:11:28 [   99.212912] testsuite-70.sh[4481]: ==4481==Hint: address points to the zero page.
23:11:28 [   99.790571] testsuite-70.sh[4481]:     #0 0x0  (<unknown module>)
23:11:28 [   99.791741] testsuite-70.sh[4481]:     #1 0x55a2eb08ddaa in run /systemd-meson-build/../build/src/cryptenroll/cryptenroll.c:737:24
23:11:28 [   99.792406] testsuite-70.sh[4481]:     #2 0x55a2eb08ddaa in main /systemd-meson-build/../build/src/cryptenroll/cryptenroll.c:762:1
23:11:28 [   99.793220] testsuite-70.sh[4481]:     #3 0x7fa837c45ccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
23:11:28 [   99.794732] testsuite-70.sh[4481]: AddressSanitizer can not provide additional info.
23:11:28 [   99.821123] testsuite-70.sh[4481]: SUMMARY: AddressSanitizer: SEGV (<unknown module>)
23:11:28 
23:11:28 
23:11:28 Nov 06 21:58:12 H testsuite-70.sh[4481]: ==4481==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x000000000000 bp 0x7ffe228cc1f0 sp 0x7ffe228cbfb8 T0)
23:11:28 Nov 06 21:58:12 H testsuite-70.sh[4481]: ==4481==Hint: pc points to the zero page.
23:11:28 Nov 06 21:58:12 H testsuite-70.sh[4481]: ==4481==The signal is caused by a READ memory access.
23:11:28 Nov 06 21:58:12 H testsuite-70.sh[4481]: ==4481==Hint: address points to the zero page.
23:11:28 Nov 06 21:58:13 H testsuite-70.sh[4481]:     #0 0x0  (<unknown module>)
23:11:28 Nov 06 21:58:13 H testsuite-70.sh[4481]:     #1 0x55a2eb08ddaa in run /systemd-meson-build/../build/src/cryptenroll/cryptenroll.c:737:24
23:11:28 Nov 06 21:58:13 H testsuite-70.sh[4481]:     #2 0x55a2eb08ddaa in main /systemd-meson-build/../build/src/cryptenroll/cryptenroll.c:762:1
23:11:28 Nov 06 21:58:13 H testsuite-70.sh[4481]:     #3 0x7fa837c45ccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
23:11:28 Nov 06 21:58:13 H testsuite-70.sh[4481]: AddressSanitizer can not provide additional info.
23:11:28 Nov 06 21:58:13 H testsuite-70.sh[4481]: SUMMARY: AddressSanitizer: SEGV (<unknown module>)
```

poettering commented on src/cryptenroll/cryptenroll-tpm2.c:
```diff
@@ -208,25 +209,52 @@ int enroll_tpm2(struct crypt_device *cd,
+                if (rc != TSS2_RC_SUCCESS)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Could not unmarshal public key from file.");
 
```
btw, i think here we might want to check that offset macthes device_key_buffer_size to disallow trailing garbage

ddstreet replied:
ack, added

poettering commented:
Looking at the code I can#t see where the memory issue comes from...

ddstreet commented:
> Looking at the code I can#t see where the memory issue comes from...

I think the problem was I failed to call `dlopen_tpm2()` before calling `sym_Tss2_MU_TPM2B_PUBLIC_Unmarshal()`.

I also added a small commit to fix the persistent handle cleanup in `testsuite-70.cryptsetup.sh`
